File: app/src/main/java/com/example/messenger/UsersActivity.java
package com.example.messenger;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.RecyclerView;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;

import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import java.util.Random;


public class UsersActivity extends AppCompatActivity {

    private UserViewModel viewModel;
    private RecyclerView recyclerViewUsers;
    private UsersAdapter adapter;
    private FirebaseDatabase database = FirebaseDatabase.getInstance();
    private DatabaseReference reference = database.getReference("Users");
    private static final String EXTRA_CURRENT_USER_ID = "user_id";
    private String currentUserId;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_users);
        initViews();
        viewModel = new ViewModelProvider(this).get(UserViewModel.class);
        observeViewModel();
        adapter.setOnUserClickListener(otherUserId -> {
            currentUserId = getIntent().getStringExtra(EXTRA_CURRENT_USER_ID);
            Intent intent = ChatActivity
                    .newIntent(UsersActivity.this, currentUserId, otherUserId.getId());
            startActivity(intent);
        });
    }

    private void initViews() {
        recyclerViewUsers = findViewById(R.id.recyclerViewUsers);
        adapter = new UsersAdapter();
        recyclerViewUsers.setAdapter(adapter);
    }

    private void observeViewModel() {
        viewModel.getUser().observe(this, user -> {
            if(user == null) {
                Intent intent = MainActivity.newIntent(this);
                startActivity(intent);
                finish();
            }
        });
        viewModel.getUsers().observe(this, users -> adapter.setUsers(users));
    }

    public static Intent newIntent(Context context, String currentUserId) {
        Intent intent =  new Intent(context, UsersActivity.class);
        intent.putExtra(EXTRA_CURRENT_USER_ID,currentUserId);
        return intent;
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.item_menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if(item.getItemId() == R.id.itemSingOut) {
            viewModel.logout();
        }
        return super.onOptionsItemSelected(item);
    }
}

File: app/src/main/java/com/example/messenger/UsersActivity.kt
package com.example.messenger

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProvider
import com.example.messenger.adapters.UsersAdapter
import com.example.messenger.databinding.ActivityUsersBinding

class UsersActivity : AppCompatActivity() {
    private lateinit var viewModel: UserViewModel
    private lateinit var binding: ActivityUsersBinding
    private val adapter = UsersAdapter()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityUsersBinding.inflate(layoutInflater)
        setContentView(binding.root)
        viewModel = ViewModelProvider(this)[UserViewModel::class.java]
        binding.recyclerViewUsers.adapter = adapter
        listenerAdapter()
        observeViewModel()
    }

    private fun observeViewModel() {
        viewModel.getUser.observe(this) {
            if (it == null) {
                val intent = MainActivity.newIntent(this)
                startActivity(intent)
                finish()
            }

        }
        viewModel.getUsers.observe(this) { adapter.users = it }
    }

    private fun listenerAdapter() {
        adapter.onUserClickListener = UsersAdapter.OnUserClickListener {

        }
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.item_menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.itemSingOut) viewModel.logout()
        return super.onOptionsItemSelected(item)
    }

    override fun onResume() {
        super.onResume()
        viewModel.onlineUserStatus(true)
    }

    override fun onStop() {
        super.onStop()
        viewModel.onlineUserStatus(false)
    }

    companion object {
        private const val EXTRA_CURRENT_USER_ID = "user_id"

        fun newIntent(context: Context, currentUserId: String): Intent {
            val intent = Intent(context, UsersActivity::class.java)
            intent.putExtra(EXTRA_CURRENT_USER_ID, currentUserId)
            return intent
        }
    }
}